src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs

```diff
@@ -0,0 +1,566 @@
+            }
+            finally
+            {
+                foreach (MemoryHandle memoryHandle in memoryHandles)
+                {
+                    memoryHandle.Dispose();
+                }
+                pinnedSegments.Dispose();
+            }
+        }
+
+        private static unsafe ValueTask<int> WriteFileGatherAsync(SafeFileHandle handle, MemoryHandle pinnedSegments,
+            int bytesToWrite, long fileOffset, CancellationToken cancellationToken)
+        {
+            SafeFileHandle.ValueTaskSource vts = handle.GetValueTaskSource();
+            try
+            {
+                NativeOverlapped* nativeOverlapped = vts.PrepareForOperation(ReadOnlyMemory<byte>.Empty, fileOffset);
+                Debug.Assert(vts._memoryHandle.Pointer != null);
+
+                // Queue an async WriteFile operation.
+                if (Interop.Kernel32.WriteFileGather(handle, (Interop.Kernel32.FILE_SEGMENT_ELEMENT*)pinnedSegments.Pointer, bytesToWrite, IntPtr.Zero, nativeOverlapped) == 0)
+                {
+                    // The operation failed, or it's pending.
+                    int errorCode = FileStreamHelpers.GetLastWin32ErrorAndDisposeHandleIfInvalid(handle);
+                    if (errorCode == Interop.Errors.ERROR_IO_PENDING)
+                    {
+                        // Common case: IO was initiated, completion will be handled by callback.
+                        // Register for cancellation now that the operation has been initiated.
+                        vts.RegisterForCancellation(cancellationToken);
+                    }
+                    else
+                    {
+                        // Error. Callback will not be invoked.
+                        vts.Dispose();
+                        return errorCode == Interop.Errors.ERROR_NO_DATA // EOF on a pipe. IO callback will not be called.
+                            ? ValueTask.FromResult<int>(0)
+                            : ValueTask.FromException<int>(SafeFileHandle.ValueTaskSource.GetIOError(errorCode, path: null));
+                    }
+                }
+            }
+            catch
+            {
+                vts.Dispose();
+                throw;
+            }
+
+            // Completion handled by callback.
+            vts.FinishedScheduling();
+            return new ValueTask<int>(vts, vts.Version);
+        }
+
+        private static unsafe (MemoryHandle pinnedSegments, MemoryHandle[] memoryHandles, int totalBytes) PrepareMemorySegments(IReadOnlyList<Memory<byte>> buffers)
+        {
+            // "The array must contain enough elements to store nNumberOfBytesToWrite bytes of data, and one element for the terminating NULL. "
+            Interop.Kernel32.FILE_SEGMENT_ELEMENT[] fileSegments = new Interop.Kernel32.FILE_SEGMENT_ELEMENT[buffers.Count + 1];
+            fileSegments[buffers.Count].Buffer = IntPtr.Zero;
```

Comment:
Nit: The array is zero'd.  This shouldn't be necessary.

Reply:
I know but I would prefer to keep it in case we ever change the default zeroing as we did for stack-allocated memory in System.Private.CoreLib (or someone copies the code an reuse it with other language as it's really hard to find some good, working examples of ReadFileScatter WriteFileGather APIs)

Reply:
> in case we ever change the default zeroing

We will never change the default zero'ing for `new T[]`.

That said, if you want to keep this, ok.